core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+            new_list = value_list[:idx] + [permuted_value] + value_list[idx+1:]
+            for permuted_new_list in cls._permute_index(new_list, idx + 1):
+                yield permuted_new_list
+
+    @classmethod
+    def _get_permutations_of_value(cls, value):
+        if isinstance(value, list):
+            for permuted_list in itertools.permutations(value):
+                for permuted_inner_list in cls._permute_index(
+                        list(permuted_list), 0):
+                    yield permuted_inner_list
+        else:
+            yield value
+
+    @classmethod
+    def _stringify_classified_rule(cls, rule_spec):
+        # This is based on the original
+        # exp_domain.RuleSpec.stringify_classified_rule, however it returns a
+        # list of possible matches by permuting the rule_spec inputs, since the
+        # order of a Python dict is implementation-dependent. Our stringified
+        # string may not necessarily match the one stored a long time ago in
+        # the data store.
+        if rule_spec.rule_type == exp_domain.CLASSIFIER_RULESPEC_STR:
+            yield rule_spec.rule_type
+        else:
+            rule_spec_inputs = rule_spec.inputs.values()
+            for permuted_inputs in cls._get_permutations_of_value(
+                    rule_spec_inputs):
+                param_list = [utils.to_ascii(val) for val in permuted_inputs]
+                yield '%s(%s)' % (rule_spec.rule_type, ','.join(param_list))
+
+    @classmethod
+    def _infer_which_answer_group_and_rule_match_answer(cls, state, rule_str):
+        # First, check whether it matches against the default rule, which
+        # thereby translates to the default outcome.
+        answer_groups = state.interaction.answer_groups
+        if rule_str == cls._DEFAULT_RULESPEC_STR:
+            return (len(answer_groups), 0, None)
+
+        # Otherwise, first RuleSpec instance to match is the winner. The first
+        # pass is to stringify parameters and doing a string comparison. This is
+        # efficient and works for most situations.
+        for answer_group_index, answer_group in enumerate(answer_groups):
+            rule_specs = answer_group.rule_specs
+            for rule_spec_index, rule_spec in enumerate(rule_specs):
+                possible_stringified_rules = list(
+                    cls._stringify_classified_rule(rule_spec))
+                if rule_str in possible_stringified_rules:
+                    return (answer_group_index, rule_spec_index, None)
+
+        # The second attempt involves parsing the rule string and doing an exact
+        # match on the rule parameter values. This needs to be done in the event
+        # that the Python-turned-ascii parameters have their own elements out of
+        # order (such as with a dict parameter).
+        if '(' in rule_str and rule_str[-1] == ')':
+            # http://stackoverflow.com/questions/9623114
+            unordered_lists_equal = lambda x, y: (
```

Comment:
See comments in _get_hashable_value() above -- I'm mildly concerned about sorting of inner lists.